includes/optional-modules/class-woo-member-commenting.php

```diff
@@ -0,0 +1,211 @@
+			$message .= ' ' . sprintf( __( 'If you already have a membership, then <a href="%s">sign in</a>.', 'newspack-plugin' ), $sign_in_url );
+		}
+		$message .= ' ' . self::get_purchase_membership_link();
+
+		return sprintf( '<p class="np-woo-member-commenting must-log-in">%s</p>', $message );
+	}
+
+	/**
+	 * Whether class should suppress commenting for non-members.
+	 *
+	 * @return bool True if we should suppress commenting for non-members, false otherwise.
+	 */
+	public static function should_load(): bool {
+		if ( empty( self::get_plan_slugs() ) ) {
+			return false;
+		}
+
+		return Settings::is_optional_module_active( 'woo-member-commenting' ) && function_exists( 'wc_memberships_get_user_memberships' ) && ! current_user_can( 'edit_posts' );
+	}
+
+	/**
+	 * Whether membership is required to comment for the current user.
+	 *
+	 * @param bool $skip_cache Whether to skip the cache.
+	 *
+	 * @return bool True if membership is required to comment, false otherwise.
+	 */
+	public static function require_membership_to_comment_for_user( bool $skip_cache = false ): bool {
+		// This gets called a lot – so cache the result.
+		static $require_membership_to_comment;
+		if ( $skip_cache || null !== $require_membership_to_comment ) {
+			return $require_membership_to_comment;
+		}
+
+		if ( ! self::should_load() ) {
+			$require_membership_to_comment = false;
+		} else {
+			$memberships_required_plans    = array_filter(
+				wc_memberships_get_user_memberships(),
+				fn( $membership ) => in_array( $membership->get_plan()->get_slug(), self::get_plan_slugs() )
+			);
+			$require_membership_to_comment = empty( $memberships_required_plans );
+		}
+
+		return $require_membership_to_comment;
+	}
+
+	/**
+	 * Get slugs for membership plans that should allow commenting.
+	 *
+	 * This is managed in the constant for now – note that the membership_plan_slug value in the array can be a single string or an array of strings.
+	 *
+	 * @return array Array of plan slugs that allow commenting.
+	 */
+	private static function get_plan_slugs(): array {
+		$slugs = self::get_module_setting( 'membership_plan_slug' );
+		if ( empty( $slugs ) ) {
+			return [];
+		}
+		if ( ! is_array( $slugs ) ) {
+			return [ $slugs ];
+		}
+
+		return $slugs;
+	}
+
+	/**
+	 * Get the link to purchase a membership.
+	 *
+	 * If the setting is empty, then an empty string is returned.
+	 *
+	 * @return string The link to purchase a membership or empty string.
+	 */
+	private static function get_purchase_membership_link(): string {
+		$post_id = self::get_module_setting( 'membership_purchase_post_id' );
+		if ( empty( $post_id ) || ! get_post( $post_id ) ) {
+			return '';
+		}
+
+		$post_url = get_permalink( (int) $post_id );
+		$message  = self::get_module_setting( 'membership_required_message' );
+		if ( empty( $message ) ) {
+			$message = __( 'Become a member now', 'newspack-plugin' );
+		}
+
+		return sprintf( '<a href="%s">%s</a>.', $post_url, $message );
```

Comment:
Non-blocking because I'm unsure: Not sure if we should include punctuation here, in case the message contains its own punctuation ("Become a member today!")

```suggestion
		return sprintf( '<a href="%s">%s</a>', $post_url, $message );
```

Reply:
Agreed!